File: sdks/javascript/src/credentials.ts
import * as keytar from 'keytar';
import { encodePrivateKey, decodePrivateKey } from './signing';

const SERVICE_NAME = 'aim_sdk';

export interface Credentials {
  agentId: string;
  apiKey: string;
  privateKey?: Uint8Array;
}

/**
 * Store credentials securely in system keyring
 * - macOS: Keychain
 * - Windows: Credential Locker
 * - Linux: Secret Service (GNOME Keyring, KWallet)
 */
export async function storeCredentials(credentials: Credentials): Promise<void> {
  await keytar.setPassword(SERVICE_NAME, 'agent_id', credentials.agentId);
  await keytar.setPassword(SERVICE_NAME, 'api_key', credentials.apiKey);

  if (credentials.privateKey) {
    const encodedKey = encodePrivateKey(credentials.privateKey);
    await keytar.setPassword(SERVICE_NAME, 'private_key', encodedKey);
  }
}

/**
 * Load credentials from system keyring
 * Returns null if no credentials are stored
 */
export async function loadCredentials(): Promise<Credentials | null> {
  const agentId = await keytar.getPassword(SERVICE_NAME, 'agent_id');
  const apiKey = await keytar.getPassword(SERVICE_NAME, 'api_key');

  if (!agentId || !apiKey) {
    return null;
  }

  const privateKeyB64 = await keytar.getPassword(SERVICE_NAME, 'private_key');
  let privateKey: Uint8Array | undefined;

  if (privateKeyB64) {
    try {
      privateKey = decodePrivateKey(privateKeyB64);
    } catch (error) {
      // Invalid private key - continue without it
      privateKey = undefined;
    }
  }

  return {
    agentId,
    apiKey,
    privateKey,
  };
}

/**
 * Clear all stored credentials from system keyring
 */
export async function clearCredentials(): Promise<void> {
  const keys = ['agent_id', 'api_key', 'private_key', 'oauth_token'];

  for (const key of keys) {
    try {
      await keytar.deletePassword(SERVICE_NAME, key);
    } catch (error) {
      // Ignore errors for missing keys
    }
  }
}

/**
 * Check if credentials are stored in keyring
 */
export async function hasCredentials(): Promise<boolean> {
  const agentId = await keytar.getPassword(SERVICE_NAME, 'agent_id');
  return agentId !== null;
}

/**
 * Get agent ID from keyring
 */
export async function getAgentId(): Promise<string | null> {
  return await keytar.getPassword(SERVICE_NAME, 'agent_id');
}

/**
 * Get API key from keyring
 */
export async function getAPIKey(): Promise<string | null> {
  return await keytar.getPassword(SERVICE_NAME, 'api_key');
}

/**
 * Store OAuth token separately
 */
export async function storeOAuthToken(token: string): Promise<void> {
  await keytar.setPassword(SERVICE_NAME, 'oauth_token', token);
}

/**
 * Get stored OAuth token
 */
export async function getOAuthToken(): Promise<string | null> {
  return await keytar.getPassword(SERVICE_NAME, 'oauth_token');
}

/**
 * Clear OAuth token from keyring
 */
export async function clearOAuthToken(): Promise<void> {
  await keytar.deletePassword(SERVICE_NAME, 'oauth_token');
}

File: sdks/go/credentials.go
package aimsdk

import (
	"crypto/ed25519"
	"fmt"

	"github.com/zalando/go-keyring"
)

const serviceName = "aim_sdk"

// Credentials holds all stored credentials
type Credentials struct {
	AgentID    string
	APIKey     string
	PrivateKey ed25519.PrivateKey
}

// StoreCredentials saves credentials to system keyring
// On macOS: Keychain, Linux: Secret Service, Windows: Credential Locker
func StoreCredentials(creds *Credentials) error {
	// Store agent ID
	if err := keyring.Set(serviceName, "agent_id", creds.AgentID); err != nil {
		return fmt.Errorf("failed to store agent_id: %w", err)
	}

	// Store API key
	if err := keyring.Set(serviceName, "api_key", creds.APIKey); err != nil {
		return fmt.Errorf("failed to store api_key: %w", err)
	}

	// Store private key (base64 encoded)
	if len(creds.PrivateKey) > 0 {
		kp := &KeyPair{PrivateKey: creds.PrivateKey}
		encodedKey := kp.PrivateKeyBase64()
		if err := keyring.Set(serviceName, "private_key", encodedKey); err != nil {
			return fmt.Errorf("failed to store private_key: %w", err)
		}
	}

	return nil
}

// LoadCredentials retrieves credentials from system keyring
// Returns error if agent is not registered (no credentials found)
func LoadCredentials() (*Credentials, error) {
	// Load agent ID
	agentID, err := keyring.Get(serviceName, "agent_id")
	if err != nil {
		return nil, fmt.Errorf("agent not registered (no agent_id found): %w", err)
	}

	// Load API key
	apiKey, err := keyring.Get(serviceName, "api_key")
	if err != nil {
		return nil, fmt.Errorf("no api_key found: %w", err)
	}

	// Load private key (optional)
	var privateKey ed25519.PrivateKey
	privateKeyB64, err := keyring.Get(serviceName, "private_key")
	if err == nil && privateKeyB64 != "" {
		kp, err := NewKeyPairFromBase64(privateKeyB64)
		if err != nil {
			// Log warning but don't fail if private key is invalid
			privateKey = nil
		} else {
			privateKey = kp.PrivateKey
		}
	}

	return &Credentials{
		AgentID:    agentID,
		APIKey:     apiKey,
		PrivateKey: privateKey,
	}, nil
}

// ClearCredentials removes all stored credentials from system keyring
// This is useful for logout or re-registration
func ClearCredentials() error {
	keys := []string{"agent_id", "api_key", "private_key", "oauth_token"}

	var lastErr error
	for _, key := range keys {
		if err := keyring.Delete(serviceName, key); err != nil {
			// Continue deleting other keys even if one fails
			// Store last error to return
			lastErr = err
		}
	}

	return lastErr
}

// HasCredentials checks if credentials are stored in the keyring
func HasCredentials() bool {
	_, err := keyring.Get(serviceName, "agent_id")
	return err == nil
}

// GetAgentID retrieves just the agent ID from keyring
func GetAgentID() (string, error) {
	return keyring.Get(serviceName, "agent_id")
}

// GetAPIKey retrieves just the API key from keyring
func GetAPIKey() (string, error) {
	return keyring.Get(serviceName, "api_key")
}

// StoreOAuthToken stores OAuth token separately
func StoreOAuthToken(token string) error {
	return keyring.Set(serviceName, "oauth_token", token)
}

// GetOAuthToken retrieves stored OAuth token
func GetOAuthToken() (string, error) {
	return keyring.Get(serviceName, "oauth_token")
}

// ClearOAuthToken removes OAuth token from keyring
func ClearOAuthToken() error {
	return keyring.Delete(serviceName, "oauth_token")
}
